### DISCUSSION TITLE
Upcoming breaking BLT changes release

### DISCUSSION THREAD

**Discussion by white238** [LLNL/blt#692] [Announcements]
BLT and CMake has grown a lot since BLT's creation. It was always the intention to avoid breaking changes but it is time to mature along with CMake. The intention of this is to discuss and socialize changes.

- Remove `BLT_DEFINES`
- Remove BLT language flags: `BLT_C_FLAGS`, `BLT_CXX_FLAGS`, `BLT_CUDA_FLAGS`, and `BLT_FORTRAN_FLAGS`
- Remove `BLT_EXE_LINKER_FLAGS`
- Use `GNUInstallDirs` and remove creating generic paths [here](https://github.com/LLNL/blt/blob/f01e0627c617bc658e52632f0922ffbcd6f4bd58/SetupBLT.cmake#L153-L186)
- Remove `CUDA_LINK_WITH_NVCC` and `CUDA_SEPARABLE_COMPILATION` (note: these are non-prefixed BLT variables)
- Remove `blt_register_library()`
- Remove `BLT_EXPORT_THIRDPARTY` and `blt_export_tpl_targets()`
- Guard all BLT options with `BLT_`: `ENABLE_CUDA`, `ENABLE_TESTING`, etc.
- Remove enabling auto-detected tools: `YAPF`, `clang-format`, etc.
- Remove `TARGET` from `blt_convert_to_system_includes`

Things that need debating:
- Remove alias targets and just have `blt_` prefixed targets, ie, `cuda` and `blt::cuda` will just be `blt_cuda`
- Remove builtin TPLs (GoogleTest, GoogleBenchMark, GoogleMock, FRUIT)
- Remove `BLT_CXX_STD`

**Comment by rhornung67** [comment]
What about BLT_CXX_STD?

**Reply by white238** [reply]
I'm torn on that one, it makes sure three variables are consistent: `CMAKE_CXX_STANDARD`, `CMAKE_CUDA_STANDARD`, `CMAKE_HIP_STANDARD`. Maybe renamed? Though maybe its better to just set all three independently.

**Reply by white238** [reply]
Though maybe in BLT we just set the CUDA and HIP variables to the CXX one if the user doesn't....

**Comment by adayton1** [comment]
I think we should also prefix all the options (e.g. BLT_ENABLE_CUDA instead of ENABLE_CUDA). I've seen a lot of projects use the no-prefix style with their own variables and it gets confusing fast which variables affect which project, and there is a high potential for name conflicts.
